daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala

```diff
@@ -0,0 +1,206 @@
+      )
+      .toOption
+
+  val parseList = (s: String) => s.tail.init.split(", ").toSeq
+
+  // Converts a given SubmitError into a SubmitError. Wraps in an UnknownError if its not what we expect, wraps in a TruncatedError if we're missing resources
+  def convertStatusRuntimeException(s: StatusRuntimeException): SubmitError = {
+    import io.grpc.protobuf.StatusProto
+    import com.daml.error.utils.ErrorDetails._
+    import com.daml.error.ErrorResource
+
+    val grpcStatus = StatusProto.fromThrowable(s)
+    val details = from(grpcStatus)
+    val message = grpcStatus.getMessage()
+    val oErrorInfoDetail = details.collectFirst { case eid: ErrorInfoDetail => eid }
+    val errorCode = oErrorInfoDetail.fold("UNKNOWN")(_.errorCodeId)
+    val resourceDetails = details.collect { case rd: ResourceInfoDetail => rd }
+    val errorResourceFromString = ErrorResource.fromString _
+
+    def classNameOf[A: ClassTag]: String = implicitly[ClassTag[A]].runtimeClass.getSimpleName
+
+    // Builds an appropriate TruncatedError if the given partial function doesn't match
+    def caseErr[A <: SubmitError: ClassTag](
+        handler: PartialFunction[Seq[(ErrorResource, String)], A]
+    ): SubmitError =
+      handler
+        .lift(
+          resourceDetails.collect {
+            case ResourceInfoDetail(name, errorResourceFromString.unlift(res)) => (res, name)
+          }
+        )
+        .getOrElse(new SubmitError.TruncatedError(classNameOf[A], message))
+
+    errorCode match {
+      case "CONTRACT_NOT_FOUND" =>
+        caseErr {
+          case Seq((ErrorResource.ContractId, cid)) =>
+            SubmitError.ContractNotFound(ContractId.assertFromString(cid))
+          // TODO: what should we do here? Currently just taking the first contract id
+          case Seq((ErrorResource.ContractIds, cids)) =>
+            SubmitError.ContractNotFound(ContractId.assertFromString(parseList(cids).head))
+        }
+      case "CONTRACT_KEY_NOT_FOUND" =>
+        caseErr {
+          case Seq(
+                (ErrorResource.TemplateId, tid),
+                (ErrorResource.ContractKey, decodeValue.unlift(key)),
+              ) =>
+            SubmitError.ContractKeyNotFound(
+              GlobalKey.assertBuild(Identifier.assertFromString(tid), key)
+            )
+        }
+      case "DAML_AUTHORIZATION_ERROR" => SubmitError.AuthorizationError(message)
+      case "CONTRACT_NOT_ACTIVE" =>
+        caseErr { case Seq((ErrorResource.TemplateId, tid), (ErrorResource.ContractId, cid)) =>
+          SubmitError.ContractNotActive(
+            Identifier.assertFromString(tid),
+            ContractId.assertFromString(cid),
+          )
+        }
+      case "DISCLOSED_CONTRACT_KEY_HASHING_ERROR" =>
+        caseErr {
+          case Seq(
+                (ErrorResource.TemplateId, tid),
+                (ErrorResource.ContractId, cid),
+                (ErrorResource.ContractKey, decodeValue.unlift(key)),
```

Comment:
In this case, the behaviour is intentional - if we cannot decode the value, we assume its because it was truncated non-gracefully such that its not parsible, so the match doesn't hit and we give back the Truncated error

Reply:
fair enough.